cluster-controller/src/main/java/io/strimzi/controller/cluster/ClusterController.java

```diff
@@ -199,9 +227,17 @@ private void recreateConfigMapWatch() {
       Periodical reconciliation (in case we lost some event)
      */
     private void reconcile() {
-        kafkaClusterOperations.reconcile(namespace, labels);
-        kafkaConnectClusterOperations.reconcile(namespace, labels);
-        kafkaConnectS2IClusterOperations.reconcile(namespace, labels);
+        Set<String> namespaces;
+        if (namespace == null) {
+            namespaces = new HashSet(client.namespaces().list().getItems());
```

Comment:
This seems to be a bit strange. When you run in all namespaces, you run the reconcile from single verticle but iterate through all namespaces. When I specify multiple specific namespaces you deploy multiple verticles and than you basically in each iterate only over one namespace.

I understand that for all namespaces we actually have to reconcile all namespaces. But this seems a bit counter intuitive.  Wouldn't it be simply easier to use always only one verticle but deploy multiple watches in it? That would seem more consistent to me.

Reply:
Yes, it would be more consistent, but it would also require more watches than strictly necessary. I admit I have no idea now many namespaces a k8s installation might typically have. Maybe it's not a bit deal having multiple watchers...?

Reply:
IMHO it doesn't matter whether you have multiple verticles with one watch per verticle or one verticle with many watches. So it is more about some confusion I saw in the code (but maybe that is just me).

I guess you will find two approaches ... one big cluster with many namespaces (== one cluster per company, projects / environments are separated by namespaces) or many smaller clusters with only few namespace each (i.e. one cluster per project). 

My assumption would be that:
- In the first case nobody would want to run your cluster across all namespaces. It would either use one or only very few selected
- In the second case I do not know ... it might be either all or just one.

Special situation might be some development environments (e.g. 100 developers run each their private Kafka cluster within shared Kubernetes/OpenShift cluster). In this case the all namespaces thing might be useful (you save resources and security is not critical).

Either way, we should document this properly, because the users who choose to run in multiple namespaces would need to deal with RBAC manually.

Reply:
@scholzj I'm happy to use a watcher-per-namespace approach, but doing that raises the need to watch for the creation/deletion of namespaces (thus the deployment/undeployment of `ClusterController` verticles). Otherwise you'd have to restart the cluster controller. So on that basis, we should just watch across all namespaces (and filter on the ClusterController which ones we're actually interested in). But does that require extra authz in k8s/openshift (and more importantly how do I figure out for myself what permissions are necessary for which actions on k8s/os, so I don't have to ask you in future)?

Reply:
@tombentley I do not have problem with the single watch across all namespaces. To be honest my view was more this way:
* Have always only one ClusterController verticle
* When `namespace` is not defined, setup a single watch accross namespaces
* When one namespace is specified, create a single watch
* When multiple namespaces are specified, use single verticle but multiple watches (one per namespace)

The main reason for it is that with the global watch anyway requires you to handle more namespaces.

But your point is good ... we need to watch for namespace changes because of the reconciliation! And IMHO, we do not want to do it. Do we want t simply drop the support for all namesapces and leave only one or more namespaces? Becasue adding namespace watcher just because of this seems overkill.

Reply:
Yes, I feel that it's overkill too (until such some as someone actually wants such functionality). So I will proceed with a single verticle, a singe `ClusterController` and a `Watch`-per-namespace. I suppose we could have a timer per namespace too, which should help isolate the namespaces a little more.

Reply:
That sounds fine to me ... although to be honest ... if we ditch all namespaces it would also remove the part which I didn't like and I would have no more issues about the original approach with verticle per namespace ;-). So I leave it up to you.